### PR TITLE
[rustdoc] Do not consider nested functions as main function even if named `main` in doctests

### LINKED ISSUE #131893: "the main function must be defined at the crate level" for doc test regression 1.81->1.82
The tokio.rs website tests all the code blocks in markdown files by building up a rust file with code included as doc comments on generated functions. The code is generally wrapped in a function to avoid any main functions being interpreted as a test main function. In 1.81, this worked fine, but 1.82 sees the internal main function and treats it as one which prevents the default test main being created.

### Code

I tried this code:

````rust
/// ```
/// # fn dox() {
/// fn main() {}
/// # }
/// ```
fn foo() {}
````

I expected to see this happen: `cargo test` should succeed

Instead, this happened: cargo test failed

### Version it worked on

It most recently worked on: 1.81

`rustc --version --verbose`:
```
rustc 1.82.0 (f6e511eec 2024-10-15)
binary: rustc
commit-hash: f6e511eec7342f59a25f7c0534f1dbea00d01b14
commit-date: 2024-10-15
host: aarch64-apple-darwin
release: 1.82.0
LLVM version: 19.1.1
```

@rustbot modify labels: +regression-from-stable-to-stable -regression-untriaged

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #131893.

If a nested function is called `main`, it is not considered as the entry point of the program. Therefore, doctests should not consider such functions as such either.

r? @notriddle

### CODE REVIEW

File: src/librustdoc/doctest/make.rs [hunk 1]
```diff
@@ -298,12 +303,14 @@ fn parse_source(
         match item.kind {
             ast::ItemKind::Fn(ref fn_item) if !info.has_main_fn => {
                 if item.ident.name == sym::main {
-                    info.has_main_fn = true;
+                    info.has_main_fn = is_top_level;
```
Comment (notriddle) [inline src/librustdoc/doctest/make.rs:306]:
```suggestion
                    info.has_main_fn = info.has_main_fn || is_top_level;
```

Also, please add this test case

````
/// ```
/// fn main() {
///     fn main() {}
/// }
/// ```
````

Comment (GuillaumeGomez) [reply src/librustdoc/doctest/make.rs:306]: Oof, good catch!

**Comment by jieyouxu** [comment]
Something about the queue is really messed up, I'll do a resync
@bors treeclosed=1000

**Comment by jieyouxu** [comment]
??? Does editing the treeclosed comment reopen the tree
@bors treeclosed=1000

**Comment by jieyouxu** [comment]
It's as if this PR is *still* being tested, so to be safe:
@bors r- retry

**Comment by jieyouxu** [comment]
Seems to have finished resyncing, so
@bors treeclosed-

**Comment by GuillaumeGomez** [comment]
Should we backport it?

**Comment by notriddle** [comment]
> Should we backport it?

I wouldn't. There's a very easy workaround, it went a long time without anyone hitting the bug, and it causes a failed compile instead of something worse (like a miscompile or a spurious test pass).

**Pull request merged**